File: packages/sdk-typescript/src/tests/conformance/ws-subscriptions.test.ts
import test from "node:test";
import assert from "node:assert/strict";

import { GeminiWebSocket } from "../../websocket/server.js";
import type {
  DepthUpdatesOptions,
  PartialDepthOptions,
  WebSocketStream,
} from "../../websocket/public.js";
import type {
  WebSocketAccountIntervalOptions,
  WebSocketScopeOptions,
} from "../../websocket/server.js";
import type { AuthStrategy } from "../../transport/http.js";
import { parseBoundaryRecord } from "../support/http-fixtures.js";
import { createWebSocketHarness } from "../support/ws-harness.js";
import type { BoundaryRecord, BoundaryValue } from "../../utils/boundary-value.js";
import { isBoundaryObject } from "../../utils/boundary-value.js";
import { loadCase, loadManifest } from "./support/fixtures.js";

async function flush(): Promise<void> {
  await new Promise<void>((resolve) => setImmediate(resolve));
}

function auth(): AuthStrategy {
  return {
    nextNonce: () => "1700000000",
    credentialHeaders: async (payloadBase64) => ({
      "X-GEMINI-APIKEY": "conformance-api-key",
      "X-GEMINI-SIGNATURE": `conformance:${payloadBase64}`,
    }),
  };
}

function record(value: BoundaryValue | undefined): BoundaryRecord | undefined {
  return isBoundaryObject(value) ? value : undefined;
}

function subscribe(client: GeminiWebSocket, fixture: BoundaryRecord): WebSocketStream<BoundaryValue> {
  const stream = String(fixture.stream);
  const symbol = typeof fixture.symbol === "string" ? fixture.symbol : undefined;
  const options = record(fixture.options);
  switch (stream) {
    case "trades":
      assert.equal(typeof symbol, "string");
      return client.public.trades(symbol!) as unknown as WebSocketStream<BoundaryValue>;
    case "bookTicker":
      assert.equal(typeof symbol, "string");
      return client.public.bookTicker(symbol!) as unknown as WebSocketStream<BoundaryValue>;
    case "depthUpdates":
      assert.equal(typeof symbol, "string");
      return client.public.depthUpdates(symbol!, options as unknown as DepthUpdatesOptions) as unknown as WebSocketStream<BoundaryValue>;
    case "partialDepth":
      assert.equal(typeof symbol, "string");
      return client.public.depth(symbol!, options as unknown as PartialDepthOptions) as unknown as WebSocketStream<BoundaryValue>;
    case "contractStatus":
      return client.public.contractStatus() as unknown as WebSocketStream<BoundaryValue>;
    case "orders":
      return client.private.orders(options as unknown as WebSocketScopeOptions) as unknown as WebSocketStream<BoundaryValue>;
    case "balances":
      return client.private.balances(options as unknown as WebSocketAccountIntervalOptions) as unknown as WebSocketStream<BoundaryValue>;
    case "positions":
      return client.private.positions(options as unknown as WebSocketAccountIntervalOptions) as unknown as WebSocketStream<BoundaryValue>;
    default:
      assert.fail(`unsupported wsSubscription stream: ${stream}`);
  }
}

function expectedParams(value: BoundaryValue): string[] {
  assert.ok(Array.isArray(value), "subscription fixture expect.params must be an array");
  return value.map((item) => {
    assert.equal(typeof item, "string");
    return item as string;
  });
}

test("wsSubscription fixtures emit exact subscribe frames", async () => {
  const manifest = await loadManifest();
  const suite = manifest.suites.find((candidate) => candidate.kind === "wsSubscription");
  assert.ok(suite, "manifest must declare a wsSubscription suite");

  for (const caseId of suite.cases) {
    await test(caseId, async () => {
      const fixture = await loadCase(suite.id, caseId);
      assert.equal(fixture.kind, "wsSubscription");
      const needsAuth = ["orders", "balances", "positions"].includes(String(fixture.stream));
      const harness = createWebSocketHarness();
      const client = new GeminiWebSocket({
        url: "wss://example.test",
        auth: needsAuth ? auth() : undefined,
        socketFactory: harness.socketFactory,
      });
      try {
        const stream = subscribe(client, fixture);
        await flush();
        assert.equal(harness.sockets.length, 1, "one fixture subscription must create one socket");
        const socket = harness.sockets[0]!;
        socket.fireOpen();
        await flush();

        const sent = socket.sent
          .map((frame) => parseBoundaryRecord(frame))
          .filter((frame) => frame.method === "SUBSCRIBE");
        assert.equal(sent.length, 1, "one fixture subscription must emit one SUBSCRIBE frame");
        const frame = sent[0]!;
        const expected = record(fixture.expect);
        assert.ok(expected);
        assert.equal(frame.method, expected.method);
        assert.deepEqual(frame.params, expectedParams(expected.params));
        assert.equal(typeof frame.id, "number");
        assert.ok(Number.isSafeInteger(frame.id) && frame.id > 0, "subscription IDs must be positive integers");

        socket.fireMessage({ data: `{"id":${String(frame.id)},"status":200}` });
        await stream.ready;
      } finally {
        client.close();
      }
    });
  }
});

File: packages/sdk-go/conformance/ws_subscriptions_test.go
package conformance

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"reflect"
	"sync"
	"testing"
	"time"

	"github.com/gemini/developer-platform/packages/sdk-go/auth"
	"github.com/gemini/developer-platform/packages/sdk-go/websocket"
)

// conformanceWSConn is an in-memory WebSocket transport. It records every
// client frame and returns the protocol acknowledgement expected by the
// subscription request path, so these tests never require a network socket.
type conformanceWSConn struct {
	mu       sync.Mutex
	closed   bool
	closedCh chan struct{}
	readCh   chan []byte
	written  [][]byte
}

func newConformanceWSConn() *conformanceWSConn {
	return &conformanceWSConn{
		closedCh: make(chan struct{}),
		readCh:   make(chan []byte, 128),
	}
}

func (c *conformanceWSConn) ReadMessage(ctx context.Context) (int, []byte, error) {
	select {
	case <-ctx.Done():
		return 0, nil, ctx.Err()
	case <-c.closedCh:
		return 0, nil, errors.New("conformance websocket connection closed")
	case payload := <-c.readCh:
		return websocket.TextMessage, payload, nil
	}
}

func (c *conformanceWSConn) WriteMessage(ctx context.Context, _ int, payload []byte) error {
	c.mu.Lock()
	if c.closed {
		c.mu.Unlock()
		return errors.New("conformance websocket connection closed")
	}
	c.written = append(c.written, append([]byte(nil), payload...))
	c.mu.Unlock()

	var request struct {
		ID     int64  `json:"id"`
		Method string `json:"method"`
	}
	if err := json.Unmarshal(payload, &request); err != nil || request.ID <= 0 || request.Method == "" {
		return nil
	}
	ack, err := json.Marshal(struct {
		ID     int64          `json:"id"`
		Status int            `json:"status"`
		Result map[string]any `json:"result"`
	}{ID: request.ID, Status: http.StatusOK, Result: map[string]any{}})
	if err != nil {
		return err
	}
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-c.closedCh:
		return errors.New("conformance websocket connection closed")
	case c.readCh <- ack:
		return nil
	}
}

func (c *conformanceWSConn) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if !c.closed {
		c.closed = true
		close(c.closedCh)
	}
	return nil
}

func (c *conformanceWSConn) feed(payload []byte) {
	select {
	case <-c.closedCh:
	case c.readCh <- append([]byte(nil), payload...):
	}
}

func (c *conformanceWSConn) frames() [][]byte {
	c.mu.Lock()
	defer c.mu.Unlock()
	frames := make([][]byte, len(c.written))
	for i, frame := range c.written {
		frames[i] = append([]byte(nil), frame...)
	}
	return frames
}

type conformanceWSDialer struct {
	conn *conformanceWSConn
}

func (d *conformanceWSDialer) Dial(context.Context, string, http.Header) (websocket.Conn, *http.Response, error) {
	return d.conn, &http.Response{StatusCode: http.StatusSwitchingProtocols}, nil
}

func TestWebSocketSubscriptionConformance(t *testing.T) {
	manifest := LoadManifest(t)
	found := false
	for _, suite := range manifest.Suites {
		if suite.Kind != "wsSubscription" {
			continue
		}
		found = true
		suiteID := suite.ID
		for _, caseID := range suite.Cases {
			caseID := caseID
			t.Run(caseID, func(t *testing.T) {
				var fixture FixtureWSSubscription
				if err := json.Unmarshal(LoadCase(t, suiteID, caseID), &fixture); err != nil {
					t.Fatalf("decode fixture: %v", err)
				}
			if fixture.Kind != "wsSubscription" {
					t.Fatalf("fixture kind = %q, want wsSubscription", fixture.Kind)
				}

				conn := newConformanceWSConn()
				dialer := &conformanceWSDialer{conn: conn}
				private := fixture.Stream == "orders" || fixture.Stream == "balances" || fixture.Stream == "positions"
				var client *websocket.Client
				if private {
					client = websocket.NewPrivateClient(
						"wss://ws.gemini.com",
						auth.NewTimeBasedHMAC(auth.APIKey("conformance-api-key"), auth.APISecret("conformance-api-secret")),
						websocket.WithDialer(dialer),
						websocket.WithAutoReconnect(false),
					)
				} else {
					client = websocket.NewPublicClient(
						"wss://ws.gemini.com",
						websocket.WithDialer(dialer),
						websocket.WithAutoReconnect(false),
					)
				}
				defer client.Close()

				ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
				defer cancel()
				if err := subscribeFixture(t, client, fixture, ctx); err != nil {
					t.Fatalf("subscribe %q: %v", fixture.Stream, err)
				}

				frames := conn.frames()
				if len(frames) != 1 {
					t.Fatalf("recorded %d wire frames, want one", len(frames))
				}
				var wire struct {
					ID     int64    `json:"id"`
					Method string   `json:"method"`
					Params []string `json:"params"`
				}
				if err := json.Unmarshal(frames[0], &wire); err != nil {
					t.Fatalf("decode SUBSCRIBE frame: %v", err)
				}
				if wire.ID <= 0 {
					t.Fatalf("SUBSCRIBE id = %d, want positive integer", wire.ID)
				}
				if wire.Method != fixture.Expect.Method {
					t.Fatalf("SUBSCRIBE method = %q, want %q", wire.Method, fixture.Expect.Method)
				}
				if !reflect.DeepEqual(wire.Params, fixture.Expect.Params) {
					t.Fatalf("SUBSCRIBE params = %#v, want %#v", wire.Params, fixture.Expect.Params)
				}
			})
		}
	}
	if !found {
		t.Fatal("manifest has no wsSubscription suite")
	}
}

func subscribeFixture(t *testing.T, client *websocket.Client, fixture FixtureWSSubscription, ctx context.Context) error {
	t.Helper()
	symbol := fixture.Symbol
	interval, err := fixtureInterval(fixture.Options)
	if err != nil {
		return err
	}
	switch fixture.Stream {
	case "trades":
		_, err = client.SubscribeTrades(ctx, symbol)
	case "bookTicker":
		_, err = client.SubscribeBookTicker(ctx, symbol)
	case "depthUpdates":
		_, err = client.SubscribeDepthWithOptions(ctx, symbol, websocket.DepthSubscriptionOptions{Interval: interval})
	case "partialDepth":
		levels, levelErr := fixtureOptionInt(fixture.Options, "levels")
		if levelErr != nil {
			return levelErr
		}
		var level websocket.PartialDepthLevel
		switch levels {
		case 0:
			level = 0
		case 5:
			level = websocket.DepthLevel5
		case 10:
			level = websocket.DepthLevel10
		case 20:
			level = websocket.DepthLevel20
		default:
			return errors.New("invalid partial-depth fixture level")
		}
		_, err = client.SubscribePartialDepth(ctx, symbol, websocket.PartialDepthSubscriptionOptions{Levels: level, Interval: interval})
	case "contractStatus":
		_, err = client.SubscribeContractStatus(ctx, symbol)
	case "orders":
		scope, scopeErr := fixtureOptionString(fixture.Options, "scope")
		if scopeErr != nil {
			return scopeErr
		}
		var subscriptionScope websocket.SubscriptionScope
		switch scope {
		case "account":
			subscriptionScope = websocket.ScopeAccount
		case "session":
			subscriptionScope = websocket.ScopeSession
		default:
			return errors.New("invalid order scope fixture")
		}
		_, err = client.SubscribeOrderEventsWithScope(ctx, subscriptionScope)
	case "balances":
		_, err = client.SubscribeBalancesWithOptions(ctx, websocket.AccountStreamOptions{Interval: interval})
	case "positions":
		_, err = client.SubscribePositionsWithOptions(ctx, websocket.AccountStreamOptions{Interval: interval})
	default:
		return errors.New("unsupported WebSocket stream fixture: " + fixture.Stream)
	}
	return err
}

func fixtureInterval(options map[string]json.RawMessage) (time.Duration, error) {
	intervalMs, err := fixtureOptionInt(options, "intervalMs")
	if err != nil {
		return 0, err
	}
	return time.Duration(intervalMs) * time.Millisecond, nil
}

func fixtureOptionInt(options map[string]json.RawMessage, key string) (int, error) {
	raw, ok := options[key]
	if !ok {
		return 0, nil
	}
	var value int
	if err := json.Unmarshal(raw, &value); err != nil {
		return 0, err
	}
	return value, nil
}

func fixtureOptionString(options map[string]json.RawMessage, key string) (string, error) {
	raw, ok := options[key]
	if !ok {
		return "", nil
	}
	var value string
	if err := json.Unmarshal(raw, &value); err != nil {
		return "", err
	}
	return value, nil
}
